From 3b864002f19fbbb7901a1431ed35bb6767a73bab Mon Sep 17 00:00:00 2001
From: Eric Yang <eric.yang2@amd.com>
Date: Tue, 24 May 2016 18:43:05 -0400
Subject: [PATCH 0513/1722] drm/amd/dal: implement voltage_state policy

only works with 1 stream currently, 2 streams
has some bug

verified use cases:
single 4k@60 splits into 2 pipes
single 5k splits into 2 pipes

Signed-off-by: Eric Yang <eric.yang2@amd.com>
Acked-by: Jordan Lazare <Jordan.Lazare@amd.com>
Signed-off-by: Alex Deucher <alexander.deucher@amd.com>
Signed-off-by: Kalyan Alle <kalyan.alle@amd.com>
---
 drivers/gpu/drm/amd/dal/dc/core/dc.c        | 38 ++++++++++++++++++++++++++---
 drivers/gpu/drm/amd/dal/dc/inc/core_types.h |  2 +-
 2 files changed, 36 insertions(+), 4 deletions(-)

diff --git a/drivers/gpu/drm/amd/dal/dc/core/dc.c b/drivers/gpu/drm/amd/dal/dc/core/dc.c
index ae9d4d0..0ca1d5a 100644
--- a/drivers/gpu/drm/amd/dal/dc/core/dc.c
+++ b/drivers/gpu/drm/amd/dal/dc/core/dc.c
@@ -872,6 +872,9 @@ bool dc_commit_surfaces_to_target(
 	struct core_target *target = DC_TARGET_TO_CORE(dc_target);
 	struct dc_target_status *target_status = NULL;
 	struct validate_context *context;
+#ifdef DIAGS_BUILD
+	struct validate_context *temp_context;
+#endif
 	int current_enabled_surface_count = 0;
 	int new_enabled_surface_count = 0;
 	bool is_mpo_turning_on = false;
@@ -949,9 +952,21 @@ bool dc_commit_surfaces_to_target(
 		goto unexpected_fail;
 	}
 
-	if (core_dc->res_pool.funcs->map_vmin_resources) {
-		core_dc->res_pool.funcs->map_vmin_resources(core_dc, context);
-	}
+#ifdef DIAGS_BUILD
+        if (core_dc->res_pool->funcs->apply_clk_constraints) {
+                temp_context = core_dc->res_pool->funcs->apply_clk_constraints(
+                                core_dc,
+                                context);
+                if (!temp_context) {
+                        dm_error("%s:failed apply clk constraints\n", __func__);
+                        BREAK_TO_DEBUGGER();
+                        goto unexpected_fail;
+                }
+                resource_validate_ctx_destruct(context);
+                dm_free(context);
+                context = temp_context;
+        }
+#endif
 
 	if (prev_disp_clk < context->bw_results.dispclk_khz ||
 		(is_mpo_turning_on &&
@@ -1003,6 +1018,23 @@ bool dc_commit_surfaces_to_target(
 					dc_surface->dst_rect.width,
 					dc_surface->dst_rect.height);
 
+
+			dal_logger_write(core_dc->ctx->logger,
+					LOG_MAJOR_HW_TRACE,
+					LOG_MINOR_HW_TRACE_SET_MODE,
+					"Pipe %d: width, height, x, y\n"
+					"viewport:%d, %d, %d, %d\n"
+					"recout:  %d, %d, %d, %d\n",
+					pipe_ctx->pipe_idx,
+					pipe_ctx->scl_data.viewport.width,
+					pipe_ctx->scl_data.viewport.height,
+					pipe_ctx->scl_data.viewport.x,
+					pipe_ctx->scl_data.viewport.y,
+					pipe_ctx->scl_data.recout.width,
+					pipe_ctx->scl_data.recout.height,
+					pipe_ctx->scl_data.recout.x,
+					pipe_ctx->scl_data.recout.y);
+
 			if (!pipe_ctx->surface->public.flip_immediate)
 				lock_mask |= PIPE_LOCK_CONTROL_SURFACE;
 
diff --git a/drivers/gpu/drm/amd/dal/dc/inc/core_types.h b/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
index ef0bb23..f6865c8 100644
--- a/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
+++ b/drivers/gpu/drm/amd/dal/dc/inc/core_types.h
@@ -234,7 +234,7 @@ struct resource_funcs {
 					const struct core_dc *dc,
 					struct validate_context *context);
 
-	enum dc_status (*map_vmin_resources)(
+	struct validate_context *(*apply_clk_constraints)(
 					const struct core_dc *dc,
 					struct validate_context *context);
 };
-- 
2.7.4

